docs/RFCS/cluster_upgrade_tool.md

```diff
@@ -0,0 +1,107 @@
+- Feature Name: Cluster Upgrade Tool
+- Status: draft
+- Start Date: 2016-09-15
+- Authors: Daniel Harrison
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: (one or more # from the issue tracker)
+
+# Summary
+
+A series of hooks to perform any necessary bookkeeping before a CockroachDB
+version upgrade.
+
+# Motivation
+
+Backward compatibility is crucial to a database, but can introduce considerable
+complexity to a codebase. All else being equal, it's better to keep backward
+compatibility logic siloed from the significant complexity that is inherent in
+software as complex as CockroachDB.
+
+# Detailed design
+
+Some versions of CockroachDB will require that a "system migration hook" be run
+before the cluster is upgraded to that version. This is expected to be
+infrequent; not every release will bump the version, only ones that need a
+migration hook.
+
+Migration hooks consist of a name, a work function, and optionally a minimum
+version. In the steady state, there is a map in code between migration names and
+the cockroach version they were released in. There is also a list of migrations
+added since the last cockroach release.
+
+When a node starts up, it checks that all known migrations have been run (the
+amount of work involved in this can be kept bounded using the `/SystemVersion`
+described below). If not, it runs them via a migration coordinator. (We should
+also do this check when a node rejoins a cluster it's been partitioned from, but
+I'm not sure how to do that.)
```

Comment:
Why? As long as the migration process completed, we know we won't regress. It may be worth repeating the check periodically to see if the node that started the migration process died without finishing it, but I don't see any reason to tie that to rejoining after a partition.